sklearn/metrics/_classification.py

```diff
@@ -320,6 +330,19 @@ def confusion_matrix(
     if y_type not in ("binary", "multiclass"):
         raise ValueError("%s is not supported" % y_type)
 
+    if y_true.size == 0 and y_pred.size == 0:
+        # early return for empty arrays avoiding all checks
+        n_classes = 0 if labels is None else len(labels)
+        return np.zeros((n_classes, n_classes), dtype=int)
+
+    if y_type == "binary":
+        pos_label = _check_pos_label_consistency(pos_label, y_true)
+    elif pos_label is not None:
```

Comment:
I am wondering if we are not shooting in our own foot with this `ValueError`.

If we change the default `pos_label=1` then we will by default raise this error whenever `y_type != "binary"`. If we choose, `pos_label=1` as a default, then we need to have the same strategy than the precision recall meaning that we need to ignore `pos_label`.

Am I right @lucyleeow or I missed something?

Reply:
Another advantages to implicitely ignoring `pos_label` is that we don't need anymore to call `_check_target` and thus the unique function because we can pass the default value most of the time.

Reply:
> hen we will by default raise this error whenever y_type != "binary"

I think we could set default to 1 and change this to:
```suggestion
    elif pos_label != 1:
```

`_check_set_wise_labels` does similar checking:

https://github.com/scikit-learn/scikit-learn/blob/e5178c53e6c81fa98d974cddbf1d79788a9885bd/sklearn/metrics/_classification.py#L1549-L1556

the `precision_recall_fscore_support` functions use it and default is `pos_label=1` (`None` option is allowed byt not documented)

(I may have gotten something wrong, this stuff gets confusing)

Reply:
I realise you could technically just flip the labels by using the 'reorder' functionality of `labels` though it is less clear and not consistent with other places where we use `pos_label` for binary case.

Reply:
>  'reorder' functionality of labels

Indeed but once I found this out, I was surprise because the API is kind of not consistent. I find it a bit better to have consistency across metrics if possible.

> I think we could set default to 1 and change this to:

The problem is that you let the case where someone pass `pos_label=1` explicitly in case other than "binary". So if we skip this test, I would rather prefer to be lenient as in other metrics.

Reply:
> The problem is that you let the case where someone pass pos_label=1 explicitly in case other than "binary".

Yes true, this is an edge case.

>  I would rather prefer to be lenient as in other metrics.

Sorry can you explain? I think with `precision_recall_fscore_support` the warning will not be raised with the above edge case either?

Reply:
> the warning will not be raised with the above edge case either?

Which is not to say it is the right way. Maybe what we want is to support the use of `None` when it means `pos_label` is to be ignored?

Reply:
> Sorry can you explain? I think with precision_recall_fscore_support the warning will not be raised with the above edge case either?

I missed this message. Imagine that you have by default, `pos_label=1`, then a call

`precision_recall_fscore_support([1, 2, 3], [1, 2, 3])` will raise an error because in this case we require `pos_label=None`.

Reply:
> To me there is two solutions:

Yes agree, these seem to be the only options. I guess I prefer the 2nd. It would be in line with what we decided in https://github.com/scikit-learn/scikit-learn/issues/10010

Reply:
> Or make 1 the default and completely ignore it in cases other than binary classification

When you say 'completely ignore', do you mean you would not raise an error if `pos_label` is not the default value and it is not the binary case?

Reply:
> do you mean you would not raise an error if pos_label is not the default value and it is not the binary case?

Yes, no raising any error.

Reply:
Options:

1. default is `None` and mention that it is equivalent to 1 in binary case
2. default is `1` and completely ignore it in cases other than binary classification (do not raise error/warning)
3. default is `1`, in case off binary classification, raise warning if `pos_label` is not the default

Now I do not know what my preference would be. I would think that it is okay to raise a warning when `pos_label` is not default but target is not binary. We do miss the case when the user wants `pos_label` to be default. (1) has the problem of `pos_label=None` being ill-defined.

But maybe more opinions are needed here. I will try to summarise in https://github.com/scikit-learn/scikit-learn/issues/10010

Reply:
Thanks. I will try to ping more dev in #10010 such that we can settle on a solution.

Reply:
Also, when `y` is a dataframe or `series` we should use `y.unique` to make this more efficient (pandas does not need to sort the data to compute the unique values).